### DISCUSSION TITLE
Authentication with Service Principal (Azure App Registration)

### DISCUSSION THREAD

**Discussion by princzkeljanos** [microsoft/Dual-write-automations#37] [Ideas]
Hi,

I see in the code that there was a trial to enable authentication with the service principal. Do you know if there is a plan to finish this code?
I have tried to finish it (the authentication was not working; it did not return the token for authentication). Unfortunately, getting over this issue led to another one where I got stuck.

In "DWEnvCalls" the "getEnvironment" method tries to get environment details, but "baseURL" in "GlobalVars" is null :-(
This is supposed to be set in the "EdgeUniversal" class in the "NetworkAdapter_ResponseReceived" event handler, where it catches responses with some nasty conditions. Of course, this part is not called when using App Registrations (Service Principal), and the base URL stays "null". This base URL is supposed to be something like this: "https://projectmanagementservice.{region}-{some id}.gateway.prod.island.powerapps.com". I could not figure out how to retrieve this to get things working with the Service Principal (App registration) authentication.

I would need this because we have strict IT, and running things in pipelines with user accounts is not welcome here. The IT would be happier using a Service Principal (App Registration).

Thanks,
Jan

**Comment by princzkeljanos** [comment]
@Hax1337 I have tried pushing my changes to a separate branch and creating a pull request for you to approve. I fixed several bugs regarding App Registration authentication. Can I get my changes to you to improve this tool?

**Comment by Hax1337** [comment]
hey @princzkeljanos yea just checkin to a seperate branch that should do, Ill check it

**Reply by princzkeljanos** [reply]
Hi, @Hax1337 I am getting the following error when I try to push my branch named "fixServicePrincipalLogin"

Pushing fixServicePrincipalLogin
Remote: Permission to microsoft/Dual-write-automations.git denied to princzkeljanos.
Error encountered while pushing to the remote repository: Git failed with a fatal error.
Git failed with a fatal error.
unable to access 'https://github.com/microsoft/Dual-write-automations/': The requested URL returned error: 403

**Comment by Hax1337** [comment]
@princzkeljanos  App registrations are not working for the dual-write page itself. There is a way to use it, but it still uses user_imposonisation. 

The code works already but it requires some setup on Azure. it is similar to the LCS authentication and similar to that setup where you need to pass username / password into it.

**Reply by princzkeljanos** [reply]
@Hax1337, can you provide any help on what should be done in Azure and how to parameterize the command to use the client app? Our IT has serious problems with passing user passwords :-(

Do I understand correctly that even if I use a client app, I have to pass the password of the user account? As I saw in the code, the secret is used in the password parameter when logging in as a client app. It is not a problem if I have to pass the username beside the secret, and it acts in the user's name; the point is not to send the user password.

Thanks.

**Reply by Hax1337** [reply]
That is correct, you still need to pass a password. Unfortuntaley there is not much I can do as the limitation is coming from the backend.
